lib/reversetunnel/agent_test.go

```diff
@@ -18,89 +18,251 @@ package reversetunnel
+}
+
+func (m *mockSSHChannel) Read(data []byte) (int, error) {
+	return 0, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Write(data []byte) (int, error) {
+	return 0, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Close() error { return nil }
+
+func (m *mockSSHChannel) CloseWrite() error { return nil }
+
+func (m *mockSSHChannel) SendRequest(name string, wantReply bool, payload []byte) (bool, error) {
+	if m.MockSendRequest != nil {
+		return m.MockSendRequest(name, wantReply, payload)
+	}
+
+	return false, trace.NotImplemented("")
+}
+
+func (m *mockSSHChannel) Stderr() io.ReadWriter {
+	return nil
+}
 
-	_, err = agent.connect()
-	require.Error(t, err, "agent should reject invalid host certificate")
+// mockAgentInjection impements several interaces for injecting into an agent.
+type mockAgentInjection struct {
+	client SSHClient
 }
 
-type fakeClient struct {
-	auth.Client
-	caKey ssh.PublicKey
+func (m *mockAgentInjection) transport(context.Context, ssh.Channel, <-chan *ssh.Request, ssh.Conn) *transport {
+	return &transport{}
 }
 
-func (fc *fakeClient) GetCertAuthorities(ctx context.Context, caType types.CertAuthType, loadKeys bool, opts ...services.MarshalOption) ([]types.CertAuthority, error) {
-	ca, err := types.NewCertAuthority(types.CertAuthoritySpecV2{
-		Type:         types.HostCA,
-		ClusterName:  "example.com",
-		CheckingKeys: [][]byte{ssh.MarshalAuthorizedKey(fc.caKey)},
+func (m *mockAgentInjection) DialContext(context.Context, utils.NetAddr) (SSHClient, error) {
+	return m.client, nil
+}
+
+func testAgent(t *testing.T) (*Agent, *mockSSHClient) {
+	tracker, err := track.New(context.Background(), track.Config{
+		ClusterName: "test",
 	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	require.NoError(t, err)
+
+	addr := utils.NetAddr{Addr: "test-proxy-addr"}
+	tracker.Start()
+
+	lease := <-tracker.Acquire()
+
+	client := &mockSSHClient{
+		MockPrincipals:        []string{"default"},
+		MockGlobalRequests:    make(chan *ssh.Request),
+		MockHandleChannelOpen: make(chan ssh.NewChannel),
+	}
+
+	inject := &mockAgentInjection{
+		client: client,
 	}
-	return []types.CertAuthority{ca}, nil
+
+	agent, err := NewAgent(&agentConfig{
+		keepAlive:   time.Minute,
+		addr:        addr,
+		transporter: inject,
+		sshDialer:   inject,
+		tracker:     tracker,
+		lease:       lease,
+	})
+	require.NoError(t, err, "Unexpected error during agent construction.")
+
+	return agent, client
 }
 
-func (fc *fakeClient) GetClusterNetworkingConfig(ctx context.Context, opts ...services.MarshalOption) (types.ClusterNetworkingConfig, error) {
-	return types.NewClusterNetworkingConfigFromConfigFile(types.ClusterNetworkingConfigSpecV2{})
+// TestAgentFailedToClaimLease tests that an agent fails when it cannot claim a lease.
+func TestAgentFailedToClaimLease(t *testing.T) {
+	agent, client := testAgent(t)
+	claimedProxy := "claimed-proxy"
+
+	var calls int
+	agent.stateCallback = func(a *Agent) {
+		calls++
+	}
```

Comment:
is there any chance of a race being introduce with this pattern? From what I can tell `stateCallback` is called in a goroutine and at that point the agent mutex is released which could depending on the scheduler lead to concurrent calls of the callback.

Reply:
Fixed races here [4305244](https://github.com/gravitational/teleport/pull/10444/commits/4305244b656ca080575a7049b3f72a547b8c134f)